### PR TITLE
Denormalize contract creation address on transaction

### LINKED ISSUE #430: Denormalize transactions table to add created contract address hash
We're seeing some optimization issues with having the `created_contract_address_hash` solely in the `internal_transactions` table. The easiest solution would be to add this column to the `transactions` table as well.

With a full Ethereum Mainnet Index, we will surely see optimization issues so it's best to make this change now.

### LINKED ISSUE #477: Optimize address page load
The address page has gotten slow again and queries need optimized.

### PR DESCRIPTION
**Pull request by tmecklem**
Resolves #430 
Fixes #477 

### Bugfixes
* Fixes timeout issue on address transactions list

### Enhancements
* Removed expensive left join or subselect queries that impacted the load time of the address page due to fetching the created_contract_address_hash from internal_transaction for display on the transactions cards

### Incompatible Changes
* This adds a denormalized transactions.created_contract_address_hash column

## Upgrading

* Since there's a change to the indexer and a new column, you'll need to drop and re-run the indexer:

```
mix do ecto.drop, ecto.create, ecto.migrate
```

#### Before:
```sql
EXPLAIN ANALYZE
SELECT t0."hash", t0."block_number", t0."cumulative_gas_used", t0."gas", t0."gas_price", t0."gas_used", t0."index", t0."internal_transactions_indexed_at", t0."input", t0."nonce", t0."r", t0."s", t0."status", t0."v", t0."value", t0."inserted_at", t0."updated_at", t0."block_hash", t0."from_address_hash", t0."to_address_hash", t0."created_contract_address_hash" 
FROM "transactions" AS t0 
LEFT OUTER JOIN "token_transfers" AS t1 ON t1."transaction_hash" = t0."hash" 
WHERE (((((t0."to_address_hash" = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex')) OR (t0."from_address_hash" = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex'))) OR (t0."created_contract_address_hash" = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex'))) OR (t1."to_address_hash" = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex'))) OR (t1."from_address_hash" = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex'))) 
ORDER BY t0."block_number" DESC, t0."index" DESC
LIMIT 50
```

Plan:
```
Limit  (cost=1000.61..12682.76 rows=50 width=403) (actual time=792.794..50265.639 rows=50 loops=1)
  ->  Gather Merge  (cost=1000.61..24636318.75 rows=105440 width=403) (actual time=792.793..50265.624 rows=50 loops=1)
        Workers Planned: 2
        Workers Launched: 2
        ->  Nested Loop Left Join  (cost=0.58..24623148.34 rows=43933 width=403) (actual time=417.770..49175.433 rows=19 loops=3)
              Filter: ((t0.to_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (t0.from_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (t0.created_contract_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (t1.to_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (t1.from_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea))
              Rows Removed by Filter: 3373405
              ->  Parallel Index Scan using transactions_recent_collated_index on transactions t0  (cost=0.43..10119037.24 rows=4231686 width=403) (actual time=0.348..44187.337 rows=3373425 loops=3)
              ->  Index Scan using token_transfers_transaction_hash_log_index_index on token_transfers t1  (cost=0.15..3.38 rows=2 width=96) (actual time=0.001..0.001 rows=0 loops=10120274)
                    Index Cond: (transaction_hash = t0.hash)
Planning time: 120.381 ms
Execution time: 50286.224 ms
```

#### After
```sql
EXPLAIN ANALYZE
SELECT t0."hash", t0."block_number", t0."cumulative_gas_used", t0."gas", t0."gas_price", t0."gas_used", t0."index", t0."internal_transactions_indexed_at", t0."input", t0."nonce", t0."r", t0."s", t0."status", t0."v", t0."value", t0."inserted_at", t0."updated_at", t0."block_hash", t0."from_address_hash", t0."to_address_hash", t0."created_contract_address_hash" FROM "transactions" AS t0 WHERE (t0."hash" = ANY((
  SELECT t0.hash AS hash
  FROM transactions AS t0
  WHERE t0.to_address_hash = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex') OR t0.from_address_hash = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex') OR t0.created_contract_address_hash = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex')
)
UNION
(
  SELECT tt.transaction_hash AS hash
  FROM token_transfers AS tt
  WHERE tt.to_address_hash = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex') OR tt.from_address_hash = decode('9a4a90e2732f3fa4087b0bb4bf85c76d14833df1', 'hex')
)
)) 
ORDER BY t0."block_number" DESC, t0."index" DESC 
LIMIT 50
```

```
Limit  (cost=51949.98..51950.11 rows=50 width=403) (actual time=2.136..2.147 rows=50 loops=1)
  ->  Sort  (cost=51949.98..51960.43 rows=4179 width=403) (actual time=2.135..2.140 rows=50 loops=1)
        Sort Key: t0.block_number DESC, t0.index DESC
        Sort Method: quicksort  Memory: 71kB
        ->  Nested Loop  (cost=16142.77..51811.16 rows=4179 width=403) (actual time=0.477..1.548 rows=58 loops=1)
              ->  HashAggregate  (cost=16142.21..16184.00 rows=4179 width=32) (actual time=0.445..0.478 rows=58 loops=1)
                    Group Key: t0_1.hash
                    ->  Append  (cost=196.01..16131.76 rows=4179 width=32) (actual time=0.138..0.418 rows=58 loops=1)
                          ->  Bitmap Heap Scan on transactions t0_1  (cost=196.01..16074.42 rows=4175 width=33) (actual time=0.138..0.405 rows=58 loops=1)
                                Recheck Cond: ((to_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (from_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (created_contract_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea))
                                Heap Blocks: exact=51
                                ->  BitmapOr  (cost=196.01..196.01 rows=4176 width=0) (actual time=0.122..0.122 rows=0 loops=1)
                                      ->  Bitmap Index Scan on transactions_to_address_hash_index  (cost=0.00..83.80 rows=2032 width=0) (actual time=0.065..0.065 rows=3 loops=1)
                                            Index Cond: (to_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea)
                                      ->  Bitmap Index Scan on transactions_from_address_hash_index  (cost=0.00..104.63 rows=2143 width=0) (actual time=0.034..0.034 rows=55 loops=1)
                                            Index Cond: (from_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea)
                                      ->  Bitmap Index Scan on transactions_created_contract_address_hash_index  (cost=0.00..4.44 rows=1 width=0) (actual time=0.022..0.022 rows=0 loops=1)
                                            Index Cond: (created_contract_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea)
                          ->  Seq Scan on token_transfers tt  (cost=0.00..15.55 rows=4 width=32) (actual time=0.005..0.005 rows=0 loops=1)
                                Filter: ((to_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea) OR (from_address_hash = '\x9a4a90e2732f3fa4087b0bb4bf85c76d14833df1'::bytea))
              ->  Index Scan using transactions_pkey on transactions t0  (cost=0.56..8.52 rows=1 width=403) (actual time=0.017..0.017 rows=1 loops=58)
                    Index Cond: (hash = t0_1.hash)
Planning time: 7.466 ms
Execution time: 3.493 ms
```

### CODE REVIEW

File: apps/explorer/test/explorer/import_test.exs [hunk 1]
```diff
@@ -113,5 +113,141 @@ defmodule Explorer.Chain.ImportTest do
...
+              value: 0
+            }
+          ]
+        }
+      }
+
+      assert {:ok, _} = Import.all(options)
+
+      transaction =
+        Explorer.Repo.one(from(transaction in Explorer.Chain.Transaction, where: transaction.hash == ^transaction_hash))
```
Comment (feliperenan) [inline apps/explorer/test/explorer/import_test.exs]: Another option would be using the function [get/3](https://hexdocs.pm/ecto/Ecto.Repo.html#c:get/3).

```Elixir
Explorer.Repo.get(Transaction, trasaction_hash)
```

File: apps/explorer/test/explorer/import_test.exs [hunk 2]
```diff
@@ -113,5 +113,141 @@ defmodule Explorer.Chain.ImportTest do
...
+
+      refute transaction.created_contract_address_hash == nil
```
Comment (feliperenan) [inline apps/explorer/test/explorer/import_test.exs]: We also could assert if the `created_contract_address_hash` is the one that we expected. Wdyt?

File: apps/explorer/test/explorer/import_test.exs (hunk 1 above)
Comment (Stamates) [reply apps/explorer/test/explorer/import_test.exs]: 👍

File: apps/explorer/test/explorer/import_test.exs (hunk 2 above)
Comment (Stamates) [reply apps/explorer/test/explorer/import_test.exs]: 👍

File: apps/explorer/lib/explorer/chain/import.ex [hunk 3]
```diff
@@ -618,7 +618,17 @@ defmodule Explorer.Chain.Import do
       from(
         t in Transaction,
         where: t.hash in ^ordered_transaction_hashes,
-        update: [set: [internal_transactions_indexed_at: ^timestamps.updated_at]]
+        update: [
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/import.ex]: The function name `update_transactions_internal_transactions_indexed_at` should be renamed as it no longer only updates the `internal_transactions_indexed_at`.

File: apps/explorer/lib/explorer/etherscan.ex [hunk 4]
```diff
@@ -45,38 +45,37 @@ defmodule Explorer.Etherscan do
   end
 
   @transaction_fields [
-    :block_number,
-    :hash,
-    :nonce,
     :block_hash,
-    :index,
+    :block_number,
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/etherscan.ex]: ❤️ keeping them sorted

File: apps/explorer_web/test/explorer_web/features/viewing_blocks_test.exs [hunk 5]
```diff
@@ -76,12 +76,18 @@ defmodule ExplorerWeb.ViewingBlocksTest do
   test "contract creation is shown for to_address in transaction list", %{session: session} do
     block = insert(:block, number: 42)
 
+    contract_address = insert(:contract_address)
+
     transaction =
       :transaction
-      |> insert(to_address: nil, to_address: nil)
```
Comment (KronicDeth) [inline apps/explorer_web/test/explorer_web/features/viewing_blocks_test.exs]: That was weird

File: apps/explorer/lib/explorer/chain/import.ex [hunk 6]
```diff
@@ -604,7 +604,7 @@ defmodule Explorer.Chain.Import do
     {:ok, inserted}
   end
 
-  defp update_transactions_internal_transactions_indexed_at(internal_transactions, %{
+  defp internal_transactions_indexed_at(internal_transactions, %{
```
Comment (KronicDeth) [inline apps/explorer/lib/explorer/chain/import.ex]: I don't know if this is better as `internal_transactions_indexed_at` is a field name and the new function doesn't imply you're insert or updating the database.  `internal_transactions_indexed_at` to me would be the name of a getter/reader of the field on the `Transaction.t`.

How about `update_transactions_with_internal_transactions` or even just `update_transactions`?

That sticks with the `<action>_<table>` naming scheme like `insert_addresses`.

Comment (Stamates) [reply apps/explorer/lib/explorer/chain/import.ex]: Fine with me, you'd know best about the context of usage. 😄

Comment (Stamates) [reply apps/explorer/lib/explorer/chain/import.ex]: Changed to `update_transactions`.  Let me know if there's anything else this PR needs.

**Pull request merged**